treeherder/perf/investigation_aids.py

```diff
@@ -0,0 +1,215 @@
+
+            decision_task = self.queue.task(decision_task_id)
+            expansion = self.auth.expandScopes({"scopes": decision_task["scopes"]})
+            expression = f"in-tree:hook-action:{hook_group_id}/{hook_id}"
+
+            if not satisfiesExpression(expansion["scopes"], expression):
+                raise RuntimeError(f"Action is misconfigured: decision task's scopes do not satisfy {expression}")
+
+            result = self.hooks.triggerHook(hook_group_id, hook_id, hook_payload)
+            return result["status"]["taskId"]
+
+    @staticmethod
+    def _defaults(target, *sources):
+        """mimics lodash's defaults"""
+        for source in sources:
+            for key, value in source.items():
+                target.setdefault(key, value)
+
+        return target
+
+    @classmethod
+    def _filter_relevant_actions(cls, actions_json: dict, original_task) -> List:
+        actions = actions_json['actions']
+
+        def collect_relevant_actions(accumulating_actions, action):
+            acc_action_names = [action['name'] for action in accumulating_actions]
+
+            action_of_known_kind = cls._has_known_kind(action)
+            action_is_relevant = action['name'] not in acc_action_names
+            no_context_or_task_to_check = (not len(action['context'])) and (not original_task)
+            task_is_in_context = (original_task and original_task.get('tags') and
+                                  cls._task_in_context(action['context'], original_task['tags']))
+
+            if (action_of_known_kind and action_is_relevant and
+                    (no_context_or_task_to_check or task_is_in_context)):
+                accumulating_actions.append(action)
+            return accumulating_actions
+
+        actions = reduce(collect_relevant_actions, actions, [])
+        return list(actions)
+
+    @staticmethod
+    def _get_action(action_array: list, action_name: str) -> str:
+        try:
+            return [a for a in action_array if a["name"] == action_name][0]
+        except IndexError:
+            available_actions = ", ".join([a["name"] for a in action_array])
+            raise LookupError(f"{action_name} action is not available for this task.  Available: {available_actions}")
+
+    @classmethod
+    def _has_known_kind(cls, action):
+        return action['kind'] in cls.KNOWN_KINDS
+
+    @classmethod
+    def _task_in_context(cls, tag_set_list, task_tags):
+        return any(
+            all(task_tags.get(tag) and task_tags[tag] == tag_set[tag]
+                for tag in tag_set.keys())
+            for tag_set in tag_set_list
+        )
+
+
+class BackfillTool:
+
+    def __init__(self, taskcluster_model: TaskclusterModel):
+        self.tc_model = taskcluster_model
+
+    def backfill_job(self, job_id: str) -> str:
+        job = self._fetch_job(job_id)
+
+        if not self.can_backfill(job):
+            logger.warning(f"Cannot backfill job {job_id}.")
+            return ''
```

Comment:
I wonder if it would be better to raise an exception here? I'm not sure how this is intended to be used, but I'd expect the caller will want to propagate an error to a user.

Also, when using a sentinel value (`''` in this case) in a return value, it is very easy to miss checking it in the caller. (`None` makes a slightly better sentinel, for that reason, if an exception isn't being used)

Reply:
I was actually thinking the same about raising an exception here. Glad we have similar opinions!